Add a basic line chart - pt1: Single line with DataLayer and AxisLayer

hsunpei opened the pull request:
# Purpose
Create a simple line chart component while preserving the flexibility in reusing the common data handling logic and some presentational components across different charts.

Thus, `<DataLayer>` and `<AxisLayer>` are created to reuse the data calculations (such as domain and range computations) and the axes on the charts respectively. 

# How it works
![linechart-pt1](https://user-images.githubusercontent.com/1139698/51654101-40f5fe80-1fd1-11e9-8f51-1477006e3a60.png)

 `<ResponsiveLayer>` gets the outer width and height of the chart, and `<DataLayer>` takes following to compute the axis configurations:

- the data
- the type of data (currently `point`, `time`, and `linear`)
- the range (the inner width and inner height of the graph, which is obtained by subtracting the margin from the outer width or height)

Then the `<DataLayer>` outputs the data below:

- The `domain` of data: based on the data type and data, it computes its domain.
- Value parsing function `getValue`: the function which parses certain data like `string` into `Date` data. 
- D3 scaling function: the function which is produced from the `d3-array` library that helps us compute the position on the graph based on the value of a given data point. (**Ref:** https://d3indepth.com/scales/)

Finally, the `<LineChart>` uses the computed data from `<DataLayer>` to draw the lines and dots, and uses `<AxisLayer>` to draw the axes.

# Screenshots
The following charts show that the `<LineChart>` currently support data of three different types: `point`, `time`, and `linear`. The data used in the charts below are given as follows:
```js
const lineData = [
  { x: 0, y: 9, date: '2019-01-21 00:00:00', weekday: 'Mon' },
  { x: 1, y: 5, date: '2019-01-22 00:00:00', weekday: 'Tue' },
  { x: 2, y: 5, date: '2019-01-23 00:00:00', weekday: 'Wed' },
  { x: 3, y: 3, date: '2019-01-24 00:00:00', weekday: 'Thu'},
  { x: 4, y: 1, date: '2019-01-25 00:00:00', weekday: 'Fri' },
];
```
1. Linear data on the X-axis:
![transchart 2019-01-23 17-46-48](https://user-images.githubusercontent.com/1139698/51597854-35ee9000-1f37-11e9-9ac8-c135166d7b41.jpg)

- X-axis: variable `x`
- Y-axis: variable `y`

2. Time data on the X-axis:
![transchart 2019-01-23 17-47-15](https://user-images.githubusercontent.com/1139698/51597812-18212b00-1f37-11e9-94b2-47eda71b85dc.jpg)

- X-axis: variable `date`
- Y-axis: variable `x`

3. Point data on the X-axis:
![transchart 2019-01-23 17-47-48](https://user-images.githubusercontent.com/1139698/51597816-1a838500-1f37-11e9-9795-c1e49f4bf40c.jpg)

- X-axis: variable `weekday`
- Y-axis: variable `y`

# TODOs
- [ ] Format the tick values
- [ ] Global styling config
- [ ] Interactions with the chart such as showing tooltips
- [ ] Title on the axis
- [ ] Chart Legends
- [ ] Animations

hsunpei committed:
Add initial files for LineChart

hsunpei committed:
Integrate ResponsiveLayer into LineChart

hsunpei committed:
Add the basic props for the line chart

hsunpei committed:
Allow Pascal case for the functional components

hsunpei committed:
Prevent Prettier from changing the import syntax to double quotes

hsunpei committed:
Add the axis config calculations for the DataLayer

hsunpei committed:
Fix the double quotes

hsunpei committed:
Fix the d3 scale type definition problems

hsunpei committed:
Enfore the checkings on semi colons

hsunpei committed:
Add render function to DataLayer

hsunpei committed:
Generate the correct axis configs in DataLayer

hsunpei committed:
Fix the error which appears in other packages using `graph` package

hsunpei committed:
Make better documentations for DataLayer

hsunpei committed:
Generate a simple line using vx

hsunpei committed:
Add LinePath's type definition

hsunpei committed:
Move the d3 type definition to the global level

hsunpei committed:
Fix the issues with the render props

hsunpei committed:
Upgrade Typescript

hsunpei committed:
Add getConvertFuncFromAxis

hsunpei committed:
Add getValue to the axes to convert the data of the time scale

hsunpei committed:
Add default margins to LineChart

hsunpei committed:
Refine the documentations

hsunpei committed:
Add the left and bottom axes

hsunpei committed:
Update tslint to check to trailing commas

hsunpei committed:
Add dots on the line chart


Add index to the line chart

hsunpei committed:
Refine the documentation

hsunpei committed:
Extract the axes into a separate layer

hsunpei committed:
Refine the docs

hsunpei committed:
Fix tslint problems

hsunpei committed:
Refine the way handling styles of the axes

hsunpei committed:
Refine the chart for the docs

hsunpei committed:
Transform the line chart as SFC

hsunpei committed:
Add CHANGELOG for LineChart

hsunpei committed:
Refine the size of the graphs

hsunpei committed:
Fix tslint issues including the class methods and the yarn command

tz5514 reviewed (approved):
Awesome implementation! 🔥🔥🔥

hsunpei committed:
Improve the linter by adding tslint-eslint-rules

hsunpei committed:
Disable the unnecessary import sorting by alphabetical orders

hsunpei committed:
Rename Axis as AxisConfig

hsunpei committed:
Fix the problems encountered on Safari

Change the format of the datetime to fix the problems of date parsing encountered on Safari

hsunpei committed:
Refine the config of Prettier

hsunpei committed:
Remove the jsx-no-multiline-js rule

hsunpei committed:
Rename the variables with more precise names and extract common types

zhusee2 reviewed (approved):
Awesome integration with `vx` library!

hsunpei committed:
Extract getAxisConfig into a helper file
